### PR TITLE
FIX: Make edit categories sidebar modal work more intuitively

### PR DESCRIPTION
**Pull request by danielwaterworth**
* Load search results in displayed order so that when more categories are loaded, they appear at the end,

### CODE REVIEW

File: spec/models/category_spec.rb [hunk 1]
```diff
@@ -1550,4 +1550,17 @@
...
+  describe ".limited_categories_matching" do
+    before_all { SiteSetting.max_category_nesting = 3 }
+
+    fab!(:foo) { Fabricate(:category, name: "foo") }
+    fab!(:bar) { Fabricate(:category, name: "bar", parent_category: foo) }
+    fab!(:baz) { Fabricate(:category, name: "baz", parent_category: bar) }
+
+    it "produces results in depth-first pre-order" do
+      SiteSetting.max_category_nesting = 3
+      expect(Category.limited_categories_matching(nil, "baz").pluck(:name)).to eq(%w[foo bar baz])
```
Comment (pmusaraj) [inline spec/models/category_spec.rb:1563]: I'm sure you're thinking about this, but we need a couple more tests here:

- what happens when user searches for "bar"? I would expect the parent to be included (foo), but not the child? 
- what happens when there is no term included? I'd expect all categories to load, right?

File: app/assets/javascripts/discourse/app/models/category.js [hunk 2]
```diff
@@ -386,6 +387,28 @@ export default class Category extends RestModel {
     return data.sortBy("read_restricted");
   }
 
+  static async asyncHierarchicalSearch(term, opts) {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/models/category.js:390]: Will this endpoint only be used if there is a search term? Meaning, for the initial load of the modal, the current endpoint works OK?

File: app/assets/javascripts/discourse/app/components/sidebar/edit-navigation-menu/categories-modal.gjs [hunk 3]
```diff
@@ -307,60 +408,82 @@ export default class SidebarEditNavigationMenuCategoriesModal extends Component
...
-                  </label>
-                </div>
+                  {{/with}}
+                {{else}}
+                  <div
+                    {{didInsert this.didInsert}}
+                    data-category-level={{c.level}}
+                    class="sidebar-categories-form__category-row"
+                  >
+                    <label
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/sidebar/edit-navigation-menu/categories-modal.gjs:472]: I'm not sure this label element is needed for the "Show more" button (in the subcategory row, this ensure clicking the row toggles the checkbox).

**Review by pmusaraj — COMMENTED** [review]
I pulled this locally, overall it is looking great to me @danielwaterworth. It needs a few more tests, not too many, we can always follow up with more tests as we discover edge cases and issues.

File: app/controllers/categories_controller.rb [hunk 4]
```diff
@@ -334,6 +334,65 @@ def find
...
+      (
+        if params[:limit].present?
+          params[:limit].to_i.clamp(1, MAX_CATEGORIES_LIMIT)
+        else
+          MAX_CATEGORIES_LIMIT
+        end
+      )
+
+    categories =
+      Category
```
Comment (pmusaraj) [inline app/controllers/categories_controller.rb:367]: Need to make sure we use the guardian here and add an accompanying test, otherwise we'll leak category names.

File: spec/requests/categories_controller_spec.rb [hunk 5]
```diff
@@ -1434,4 +1434,31 @@
...
+
+    it "doesn't produce categories with a very specific term" do
+      get "/categories/hierarchical_search.json", params: { term: "acategorythatdoesnotexist" }
+
+      expect(response.status).to eq(200)
+      expect(response.parsed_body["categories"].length).to eq(0)
+    end
+
+    it "doesn't expose secret categories" do
+      category.update!(read_restricted: true)
```
Comment (pmusaraj) [inline spec/requests/categories_controller_spec.rb:1456]: Can we also test against a private category? Read restricted is a specific type of "private", but there's also a `private_category` fabricator.

File: app/controllers/categories_controller.rb [hunk 6]
```diff
@@ -334,6 +334,52 @@ def find
...
+        .limited_categories_matching(only, except, parent_category_id, term)
+        .includes(
+          :uploaded_logo,
+          :uploaded_logo_dark,
+          :uploaded_background,
+          :uploaded_background_dark,
+          :tags,
+          :tag_groups,
+          :form_templates,
+          category_required_tag_groups: :tag_group,
```
[the remark this answers is not in the document]
Comment (pmusaraj) [reply app/controllers/categories_controller.rb:378]: Any thoughts on this question @danielwaterworth? Are we loading this stuff for nothing in the modal? We'er certainly not showing logos, backgrounds or form templates.

[the remark this answers is not in the document]
Comment (danielwaterworth) [reply app/controllers/categories_controller.rb:378]: We need to use the site category serializer, since these categories get added to the site categories, so this is necessary.

**Review by pmusaraj — APPROVED** [review]
